habitat/datasets/pointnav/generator.py

```diff
@@ -0,0 +1,142 @@
+    euclid_dist = np.power(np.power(substract_vec, 2).sum(0), 0.5)
+    if np.abs(s[1] - t[1]) > 0.5:
+        return False, 0
+    d_separation = env.sim.geodesic_distance(s, t)
+    if d_separation == np.inf:
+        return False, 0
+    if not near_dist <= d_separation <= far_dist:
+        return False, 0
+    distances_ratio = d_separation / euclid_dist
+    if distances_ratio < GEODESIC_TO_EUCLID_RATIO_THRESHOLD and np.random.rand() > _ratio_sample_rate(
+        distances_ratio
+    ):
+        return False, 0
+    if env.sim.island_radius(s) < ISLAND_RADIUS_LIMIT:
+        return False, 0
+    return True, d_separation
+
+
+def _create_episode(
+    episode_id,
+    scene_id,
+    start_position,
+    start_rotation,
+    target_position,
+    shortest_paths=None,
+    info=None,
+) -> Optional[NavigationEpisode]:
+    goals = [NavigationGoal(position=target_position)]
+    return NavigationEpisode(
+        episode_id=str(episode_id),
+        goals=goals,
+        scene_id=scene_id,
+        start_position=start_position,
+        start_rotation=start_rotation,
+        shortest_paths=shortest_paths,
+        info=info,
+    )
+
+
+def get_action_shortest_path(
+    env, source_position, source_rotation, goal_position, radius=0.05
+) -> List[ShortestPathPoint]:
+    backup_episodes = env.episodes
+    if not env.episode_start_time:
+        env.reset()
+    env.episodes = [env.current_episode]
+    env.reset()
+    env.sim.set_agent_state(source_position, source_rotation)
+
+    mode = "greedy"
+
+    follower = ShortestPathFollower(env.sim, radius, False)
+    follower.mode = mode
+
+    shortest_path = []
+
+    while not env.episode_over:
+        action = follower.get_next_action(goal_position)
+        state = env.sim.get_agent_state()
+        shortest_path.append(
+            ShortestPathPoint(
+                state.position.tolist(),
+                state.rotation.components.tolist(),
+                action.value,
+            )
+        )
+        env.step(action.value)
+
+    env.episodes = backup_episodes
+    return [shortest_path]
+
+
+def generate_pointnav_episode(env, num_episodes=-1, gen_shortest_path=True):
+    episode_count = 0
+    while episode_count < num_episodes or num_episodes < 0:
+        target_position = env.sim.sample_navigable_point()
+
+        if env.sim.island_radius(target_position) < ISLAND_RADIUS_LIMIT:
+            continue
+
+        for retry in range(NUMBER_RETRIES_PER_TARGET):
+            source_position = env.sim.sample_navigable_point()
+
+            is_compatible, dist = is_compatible_episode(
+                source_position,
+                target_position,
+                env,
+                near_dist=NEAR_DIST_LIMIT,
+                far_dist=FAR_DIST_LIMIT,
+            )
+            if is_compatible:
+                angle = np.random.uniform(0, 2 * np.pi)
+                source_rotation = [0, np.sin(angle / 2), 0, np.cos(angle / 2)]
```

Comment:
Should this be `[0, np.sin(angle), 0, np.cos(angle)]` ?

Reply:
This is correct, `sin(angle / 2)` and `cos(angle / 2)` uniquely identify the `angle`. We should use a utility function for this, I think there is an `angle` to `quaternion` utility function which we can call here.

Reply:
Here is a visualization to confirm this: 
![image](https://user-images.githubusercontent.com/3365035/57318769-f3009480-70af-11e9-84cd-3097ea891ff6.png)

As the graph never intersects itself each point is uniquely identified.

Reply:
You can use something like `quat_from_angle_axis` for this https://github.com/facebookresearch/habitat-sim/blob/master/habitat_sim/utils.py#L66

In this case, the axis would be the up direction, `[0, 1, 0]` and the angle would be `theta \in [0, 2 PI]`